src/validation/rules/FieldsOnCorrectType.js

```diff
@@ -11,14 +11,70 @@
 import type { ValidationContext } from '../index';
 import { GraphQLError } from '../../error';
 import type { Field } from '../../language/ast';
-import type { GraphQLType } from '../../type/definition';
-
+import {
+  isAbstractType,
+  GraphQLAbstractType,
+  GraphQLObjectType,
+} from '../../type/definition';
 
 export function undefinedFieldMessage(
-  fieldName: string,
-  type: GraphQLType
-): string {
-  return `Cannot query field "${fieldName}" on "${type}".`;
+    field: string,
+    type: string,
+    suggestedTypes: Array<string>): string {
+  let message = `Cannot query field "${field}" on type "${type}".`;
+  const MAX_LENGTH = 5;
+  if (suggestedTypes.length !== 0) {
+    let suggestions = suggestedTypes
+      .slice(0, MAX_LENGTH)
+      .map(t => `"${t}"`)
+      .join(', ');
+    if (suggestedTypes.length > MAX_LENGTH) {
+      suggestions += `, and ${suggestedTypes.length - MAX_LENGTH} other types`;
+    }
+    message += ` However, this field exists on ${suggestions}.`;
+    message += ` Perhaps you meant to use an inline fragment?`;
+  }
+  return message;
+}
+
+/**
+ * Return implementations of `type` that include `fieldName` as a valid field.
+ */
+function getImplementationsIncludingField(
+    type: GraphQLAbstractType,
+    fieldName: string): Array<string> {
+  return type.getPossibleTypes()
+    .filter(t => t.getFields()[fieldName] !== undefined)
+    .map(t => t.name)
+    .sort();
+}
+
+/**
+ * Go through all of the implementations of type, and find other interaces
+ * that they implement. If those interfaces include `field` as a valid field,
+ * return them, sorted by how often the implementations include the other
+ * interface.
+ */
+function getSiblingInterfacesIncludingField(
+    type: GraphQLAbstractType,
+    fieldName: string): Array<string> {
+  const implementingObjects = type.getPossibleTypes()
+    .filter(t => t instanceof GraphQLObjectType);
+
+  const suggestedInterfaces = implementingObjects.reduce((acc, t) => {
+    t.getInterfaces().forEach(i => {
+      if (i.getFields()[fieldName] === undefined) {
+        return;
+      }
+      if (acc[i.name] === undefined) {
+        acc[i.name] = 0;
+      }
+      acc[i.name] += 1;
+    });
+    return acc;
+  }, {});
+  return Object.keys(suggestedInterfaces)
+    .sort((a,b) => suggestedInterfaces[b] - suggestedInterfaces[a]);
```

Comment:
Minor nit, I think it's preferable to put functions which are only used internally below any that are `export`ed, so it's easier to see what functions are exported on first read.